cancat/uds/utils.py

```diff
@@ -0,0 +1,495 @@
+# Utility functions for UDS
+
+import time
+import string
+import struct
+from contextlib import contextmanager
+
+from cancat import uds
+from cancat.uds import UDS
+from cancat.utils import log
+from cancat.utils.types import ECUAddress, _range_func
+
+def enter_session(u, session, prereq_sessions=None):
+    # Enter any required preq sessions
+    if prereq_sessions:
+        for prereq in prereq_sessions:
+            #u.DiagnosticSessionControl(prereq)
+            enter_session(u, prereq)
+
+    return u.DiagnosticSessionControl(session)
+
+@contextmanager
+def new_session(u, session, prereq_sessions=None, tester_present=False):
+    try:
+        # Enter any required preq sessions
+        if prereq_sessions:
+            for prereq in prereq_sessions:
+                #u.DiagnosticSessionControl(prereq)
+                enter_session(u, prereq)
+
+        msg = u.DiagnosticSessionControl(session)
+
+        # Start tester present again
+        if tester_present:
+            u.StartTesterPresent(request_response=False)
+
+        yield msg
+    finally:
+        if tester_present:
+            u.StopTesterPresent()
+
+
+def find_possible_resp(u, start_index, tx_arbid, service, subfunction=None, timeout=3.0):
+    # Starting at the supplied starting index, find the service request, and
+    # then look for possible responses until the supplied timeout
+    
+    if subfunction:
+        tx_match_bytes = struct.pack('>BH', service, subfunction)
+        rx_match_bytes = struct.pack('>BH', service + 0x40, subfunction)
+        match_len = 3
+    else:
+        tx_match_bytes = struct.pack('>B', service)
+        rx_match_bytes = struct.pack('>B', service + 0x40)
+        match_len = 1
+
+    for idx, _, _, msg in u.c.genCanMsgs(start=start_index, arbids=[tx_arbid], maxsecs=timeout):
+        ftype = ord(msg[0]) >> 4
+        # Our Tx arbid is probably a 0, but check for frame type 1 as well.
+        if (ftype == 0 and msg[1:1+match_len] == tx_match_bytes) or \
+                (ftype == 1 and msg[2:2+match_len] == tx_match_bytes):
+            tx_index = idx
+            tx_msg = msg
+            break
+    else:
+        err = 'Unable to find tx arbid {} starting at index {}'.format(hex(tx_arbid), start_index)
+        raise RangeError(err)
+
+    # Determine the proper rx_arbid range to look for
+    if tx_arbid > 0x18db0000:
```

Comment:
can you please make these magic numbers into CONSTANT_VALUES which are set elsewhere?

Reply:
sure